File: backend/src/acquisition/amd.js
const axios = require('axios');
const logger = require('../logger');
const { queueHttp } = require('./requestQueue');
const settings = require('../config/settings');

/**
 * Fetches jobs from AMD's career site using their new Attract/iCIMS API.
 *
 * API: GET https://careers.amd.com/api/jobs
 * Parameters:
 *   - page: page number (1-based)
 *   - limit: page size (e.g. 50)
 *   - location: e.g. "India"
 *   - keywords: search text (optional)
 *
 * @param {Object} company - Company config object
 * @returns {Array} Normalized job objects
 */
async function scrapeAmd(company) {
  const { name, filters } = company;
  const location = filters?.location || 'India';
  const keywords = filters?.keywords || '';

  logger.info(`[${name}] AMD API fetch starting`, { location, keywords });

  const jobs = [];
  let page = 1;
  const limit = 50;
  let hasMore = true;

  while (hasMore) {
    try {
      const params = new URLSearchParams({
        page: String(page),
        limit: String(limit),
        sortBy: 'relevance',
        descending: 'false',
        internal: 'false'
      });

      if (location) params.append('location', location);
      if (keywords) params.append('keywords', keywords);

      const apiUrl = `https://careers.amd.com/api/jobs?${params.toString()}`;

      const response = await queueHttp(() =>
        axios.get(apiUrl, {
          headers: {
            'User-Agent': settings.scraping.userAgent || 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
            'Accept': 'application/json',
          },
          timeout: 15000,
        })
      );

      const pageJobs = response.data?.jobs || [];

      if (pageJobs.length === 0) {
        hasMore = false;
        break;
      }

      for (const j of pageJobs) {
        const job = normalizeAmdJob(j, company);
        if (job) jobs.push(job);
      }

      logger.info(`[${name}] AMD API page ${page} fetched`, {
        fetched: pageJobs.length,
      });

      if (pageJobs.length < limit) {
        hasMore = false;
      } else {
        page++;
      }
    } catch (err) {
      const status = err?.response?.status;
      if (status === 403 || status === 401) {
        logger.error(`[${name}] Access denied (${status})`);
        throw Object.assign(err, { markDegraded: true });
      }
      logger.error(`[${name}] AMD API fetch error`, { message: err.message, status });
      hasMore = false;
    }
  }

  logger.info(`[${name}] AMD API scrape complete`, { totalJobs: jobs.length });
  return jobs;
}

function normalizeAmdJob(item, company) {
  try {
    const job = item.data;
    if (!job) return null;

    const jobId = job.req_id || job.slug || String(Date.now() + Math.random());
    const location = job.full_location || job.short_location || [job.city, job.state, job.country].filter(Boolean).join(', ');
    const jobUrl = job.meta_data?.canonical_url || job.apply_url || `https://careers.amd.com/jobs/${jobId}`;

    return {
      companyName: company.name,
      jobId: String(jobId),
      jobTitle: job.title || 'Unknown Title',
      location,
      department: (job.category && job.category[0]) || (job.categories && job.categories[0]) || '',
      postedDate: job.posted_date || job.create_date || new Date().toISOString(),
      employmentType: job.employment_type || 'Full-time',
      jobDescription: job.description || '',
      url: jobUrl,
      applyUrl: job.apply_url || jobUrl,
    };
  } catch {
    return null;
  }
}

module.exports = { scrapeAmd };

File: backend/nlp_service/adapters/amd.py
import time
import json
import re
import base64
from datetime import datetime
from utils import queue_http, HEADERS

def scrape_amd(company, filters):
    # Works for AMD
    locations = filters.get('locations') or [filters.get('location', 'India')]
    keywords = filters.get('keywords', '')

    jobs = []

    for loc in locations:
        page = 1
        limit = 50

        while True:
            params = {
                "page": str(page),
                "limit": str(limit),
                "sortBy": "relevance",
                "descending": "false",
                "internal": "false"
            }
            if loc:
                params['location'] = loc
            if keywords:
                params['keywords'] = keywords

            res = queue_http('https://careers.amd.com/api/jobs', params=params)
            if res.status_code != 200:
                res.raise_for_status()

            data = res.json()
            page_jobs = data.get('jobs', [])
            if not page_jobs:
                break

            for j in page_jobs:
                job = j.get('data', {})
                if not job:
                    continue

                job_id = job.get('req_id') or job.get('slug') or str(int(time.time()))
                loc_val = job.get('full_location') or job.get('short_location') or ", ".join(filter(None, [job.get('city'), job.get('state'), job.get('country')]))
                job_url = job.get('meta_data', {}).get('canonical_url') or job.get('apply_url') or f"https://careers.amd.com/jobs/{job_id}"

                category = (job.get('category') and job['category'][0]) or (job.get('categories') and job['categories'][0]) or ''

                jobs.append({
                    "companyName": company['name'],
                    "jobId": str(job_id),
                    "jobTitle": job.get('title', 'Unknown Title'),
                    "location": loc_val,
                    "department": category,
                    "postedDate": job.get('posted_date') or job.get('create_date') or datetime.utcnow().isoformat() + 'Z',
                    "employmentType": job.get('employment_type', 'Full-time'),
                    "jobDescription": job.get('description', ''),
                    "url": job_url,
                    "applyUrl": job.get('apply_url') or job_url
                })

            if len(page_jobs) < limit:
                break
            page += 1

    return jobs
